### PR TITLE
Hide impossible early semesters

### PR DESCRIPTION
**Pull request by whisperity**
This is yet again something you ask yourself "Why isn't this the default way it works?".

### CODE REVIEW

File: npu.user.js [hunk 1]
```diff
@@ -370,6 +370,7 @@ var npu = {
 			$("#panHeader, #panCloseHeader").hide();
 			$("table.top_menu_wrapper").css("margin-top", "5px").css("margin-bottom", "8px");
 			$("#form1 > fieldset").css("border", "0 none");
+			$("#span_changeproject").parent().hide();
```
Comment (whisperity) [inline npu.user.js:373]: Okay, I did not wish to make a separate commit for this. Long ago a new button (to the menu bar) was added for _Meet Street_, but by removing the top header bar, a "Switch to Meet Street" hyperlink is presented in the top left corner. This removes it.

File: npu.user.js [hunk 2]
```diff
@@ -506,8 +507,11 @@ var npu = {
 					var select = $('<ul class="termSelect"></ul>');
 					var stored = npu.getUserData(null, null, ["termSelect", npu.getPage()]);
 					var found = false;
+					var admissionSemester = $("#lblTrainingName").text().split("[")[0].split(":")[1];
```
Comment (solymosi) [inline npu.user.js]: This seems a bit fragile to me. What if the training name contains a "[" character, followed by a letter? Then we would be removing all semesters from the term selector as numbers come before letters.

The "subject name parsing" bugs I had to fight a while ago taught me never to underestimate the amount of weird or badly formatted data out there. What seems consistent here is that the end of the string is always in the form of `:XXXX/YY/Z[some text]`. Maybe a regex for that?

File: npu.user.js (hunk 1 above)
Comment (solymosi) [inline npu.user.js:373]: I like this :+1:

File: npu.user.js (hunk 2 above)
Comment (whisperity) [inline npu.user.js]: Having a `[` in the training's name sounds so bad, but yes, we who only live by chipping away on the clientside output may never now. I shall fix it soon.

Otherwise, you find the general idea useful?

Comment (solymosi) [inline npu.user.js]: Thanks. Otherwise I like the idea very much :+1:

File: npu.user.js [hunk 3]
```diff
@@ -506,8 +507,11 @@ var npu = {
 					var select = $('<ul class="termSelect"></ul>');
 					var stored = npu.getUserData(null, null, ["termSelect", npu.getPage()]);
 					var found = false;
+					var admissionSemester = String($("#lblTrainingName").text().match(/:(\d{4}\/\d{2}\/\d)\[.*?\]\)/)[1]);
```
Comment (solymosi) [inline npu.user.js]: Let's make this a bit more robust by adding an EOL anchor to the end:

```
/:(\d{4}\/\d{2}\/\d)\[.*?\]\)$/
```

Comment (solymosi) [inline npu.user.js]: Also, this will throw a `TypeError` if for whatever reason nothing is matched. Probably best to avoid that:

```
var match = $("#lblTrainingName").text().match(/:(\d{4}\/\d{2}\/\d)\[.*?\]\)$/);
var admissionSemester = match && String(match[1]);
```

File: npu.user.js [hunk 4]
```diff
@@ -506,8 +507,11 @@ var npu = {
 					var select = $('<ul class="termSelect"></ul>');
 					var stored = npu.getUserData(null, null, ["termSelect", npu.getPage()]);
 					var found = false;
+					var admissionSemester = String($("#lblTrainingName").text().match(/:(\d{4}\/\d{2}\/\d)\[.*?\]\)/)[1]);
+					
 					$("option", termSelect).each(function() {
 						if($(this).attr("value") == "-1") { return; }
+						if($(this).text() < admissionSemester) { return; }
```
Comment (solymosi) [inline npu.user.js]: We should prepare for the possibility that no admission semester is matched:

```
if(admissionSemester && $(this).text() < admissionSemester) { return; }
```

**Comment by whisperity** [comment]
Thanks for the reply. Going to fix them soon. :smile:

**Comment by whisperity** [comment]
Sorry, got carried away at work. :smirk:

``` js
var x = 5;
var y = (x && x + 1);
// y is 6
```

Did not believe that it works like this...

**Comment by solymosi** [comment]
Looks good, thanks :+1:
